u5-rest-apis/features/covid-controller.js

```diff
@@ -0,0 +1,134 @@
+      .docs
+      .map(doc => doc.data())
+      .map(({ date, stateId, stateName, cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent }) => ({ date: date.toMillis(), stateId, stateName, cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent }))
+
+    res.json({
+      result: 'ok',
+      payload,
+      count: payload.length
+    })
+  } catch (err) {
+    console.error(err)
+    res.status(500).json({
+      result: 'error',
+      payload: [],
+      count: 0
+    })
+  }
+}
+
+const covie = async (req, res) => {
+  try {
+    // 1. Inputs
+    const stateId = req.params.stateId.toUpperCase()
+
+    // 2. Query
+    const query = db.collection('covid-history').doc(stateId).get()
+
+    // 3. Response
+    const snapshot = await query
+    if (!snapshot.exists) return res.status(404).json({ result: 'not found' })
+
+    const data = snapshot.data()
+
+    const payload = {
+      stateId, // this is already a variable
+      stateName: data.stateName,
+      history: data.history.map(({ date, cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent }) => ({ date: date.toMillis(), cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent }))
+    }
+
+    res.json({ result: 'ok', payload })
+  } catch (err) {
+    console.error(err)
+    res.status(500).json({ result: 'error' })
+  }
+}
+
+const createCovie = async (req, res) => {
+  try {
+    // 1. Inputs
+    const { stateId, date, cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent } = req.body
+    const record = { stateId, date: firestore.Timestamp.fromMillis(date), cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent }
+
+    const historyRecord = {
+      history: firestore.FieldValue.arrayUnion({
+        date: firestore.Timestamp.fromMillis(date),
+        cases,
+        casesNew,
+        vaccineOne,
+        vaccineOnePercent,
+        vaccineComplete,
+        vaccineCompletePercent
+      })
+    }
+
+    // 2. Query
+    const query = db.collection('covid-latest').doc(stateId).set(record, { merge: true })
+    const queryhistory = db.collection('covid-history').doc(stateId).set(historyRecord, { merge: true })
+
+    // 3. Response
+    await query
+    await queryhistory
+
+    res.sendStatus(201)
+  } catch (err) {
+    console.error(err)
+    res.sendStatus(500)
+  }
+}
+
+const updateCovie = async (req, res) => {
+  try {
+    // 1. Inputs
+    const stateId = req.params.stateId.toUpperCase()
+    const { date, cases, casesNew } = req.body
+    const record = { date: firestore.Timestamp.fromMillis(date), cases, casesNew }
+
+    const historyRecord = {
+      history: firestore.FieldValue.arrayUnion({ date: firestore.Timestamp.fromMillis(date), cases, casesNew })
+    }
+    // 2. Query
+    const query = db.collection('covie-latest').doc(stateId).set(record, { merge: true })
```

Comment:
หมายเหตุ: สำหรับห้องปฏิบัติการ คุณควรถือว่า `date` ในคำขอเท่ากับ `date` ใน `covid-latest` (นี่คือการอัปเดตข้อมูล "ล่าสุด") โซลูชันที่สมจริงยิ่งขึ้นจะตรวจสอบว่าสิ่งนี้เป็นจริงโดยใช้ธุรกรรม -- ตัวอย่างเช่น การอัปเดตข้อมูลสำหรับวันที่เก่าไม่ควรเปลี่ยน `covid-latest` มันจะอัพเดทแค่ `covid-history`

Note: For the lab you should assume that `date` is equal to the `date` already in `covid-latest` (this is updating the "latest" data). A more realistic solution would check that this is true using a transaction -- for example updating data for an old date should not change `covid-latest`; it would only update `covid-history`.